src/freenet/client/filter/M3UFilter.java

```diff
@@ -0,0 +1,171 @@
+        int readcount;
+        byte[] nextchar = new byte[1];
+        byte[] fileUri;
+        int fileIndex;
+        readcount = dis.read(nextchar);
+        // read each line manually
+        while (readcount != -1) {
+            if(Arrays.equals(nextchar, CHAR_COMMENT_START)) {
+                isComment = true;
+            } else {
+                isComment = false;
+            }
+            // read one line as a fileUri
+            numberOfDotsInUri = 0;
+            isBadUri = false;
+            fileIndex = 0;
+            fileUri = new byte[MAX_URI_LENGTH];
+            while (readcount != -1) {
+                if (!isComment) {
+                    // do not include carriage return in filenames
+                    if (!Arrays.equals(nextchar, CHAR_CARRIAGE_RETURN)) {
+                        if (fileIndex <= MAX_URI_LENGTH) {
+                            fileUri[fileIndex] = nextchar[0];
+                            fileIndex += readcount;
+                        }
+                    }
+                }
+                readcount = dis.read(nextchar);
+                if (Arrays.equals(nextchar, CHAR_NEWLINE)) {
+                    if (!isComment) {
+                        // remove too long paths
+                        if (fileIndex <= MAX_URI_LENGTH) {
+                            // overly strict filtering to keep it simple for starters.
+                            // allow only alphanumeric values in UTF-8 encoding and exactly one period in the filename.
+                            final byte utf80 = (byte)'0';
+                            final byte utf89 = (byte)'9';
+                            final byte utf8A = (byte)'A';
+                            final byte utf8Z = (byte)'Z';
+                            final byte utf8a = (byte)'a';
+                            final byte utf8z = (byte)'z';
+                            final byte utf8dot = (byte)'.';
+                            final byte utf8dash = (byte)'-';
+                            for (int i = 0; i < fileIndex; i++) {
+                                byte b = fileUri[i];
+                                if (!((utf80 <= b && b <= utf89) ||
+                                      (utf8A <= b && b <= utf8Z) ||
+                                      (utf8a <= b && b <= utf8z) ||
+                                      utf8dot == b ||
+                                      utf8dash == b)) {
+                                    isBadUri = true;
+                                    break;
+                                } 
+                                if ((byte)'.' == b) {
+                                    numberOfDotsInUri += 1;
+                                }
+                            }
+                            if (numberOfDotsInUri != 1) {
+                                isBadUri = true;
+                            }
+                            // use only the first fileIndex bytes
+                            String uri;
```

Comment:
I separated it to make it clearer how the different comparisons combine: separate && and ||.